sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py

```diff
@@ -1450,6 +1450,86 @@ def test_unknown_category_that_are_negative():
     assert_allclose(hist.predict(X_test_neg), hist.predict(X_test_nan))
 
 
+@pytest.mark.parametrize(
+    ("GradientBoosting", "make_X_y"),
+    [
+        (HistGradientBoostingClassifier, make_classification),
+        (HistGradientBoostingRegressor, make_regression),
+    ],
+)
+@pytest.mark.parametrize("sample_weight", [False, True])
+def test_X_val_in_fit(GradientBoosting, make_X_y, sample_weight, global_random_seed):
+    """Test that passing X_val, y_val in fit is same as validation fraction."""
+    rng = np.random.RandomState(42)
+    n_samples = 100
+    X, y = make_X_y(n_samples=n_samples, random_state=rng)
+    if sample_weight:
+        sample_weight = np.abs(rng.normal(size=n_samples))
+        data = (X, y, sample_weight)
+    else:
+        sample_weight = None
+        data = (X, y)
+    rng_seed = global_random_seed
+
+    # Fit with validation fraction and early stopping.
+    m1 = GradientBoosting(
+        early_stopping=True,
+        validation_fraction=0.5,
+        random_state=rng_seed,
+    )
+    m1.fit(X, y, sample_weight)
```

Comment:
What should happen if `validation_fraction=0.5` and `X_val` is passed in during fit? I feel like we should error.

Reply:
I go with a warning saying that `X_val` wins over `validation_fraction`. Ok for you?

Reply:
I slightly prefer raising an error more than showing a warning.

Reply:
Why?
I prefer the warning because an error seems like patronizing. A user did not make a mistake and we should not make her/his (machine learning) life miserable.

Reply:
The default could be "auto" instead of 0.1, and then it makes sense to raise for any real value.

Reply:
> Why?

If I am reading the following without running it, it's hard to tell what the behavior is:

```python
clf = HistGradientBoosting(validation_fraction=0.3)
clf.fit(X_train, y_train, X_val=X_val, y_val=y_val)
```

> the default is validation_fraction=0.1

Ah that is unfortunate. I'm okay with the "auto" suggestion from @adrinjalali

Reply:
> If I am reading the following without running it, it's hard to tell what the behavior is

I don't think so, the docstring of validation fraction says: "It is ignored if `X_val` and `y_val` are passed to fit."

Please note that:
- `validation_fraction` is ignored if `early_stopping=False`.
- `validation_fraction=None` means to use training data for early stopping.

In light of there, I suggest the following:
- Raise an error if `X_val, y_val` are passed to `fit`, but `early_stopping=False` (default is "auto").
- Remove the warning with passed `X_val, y_val` and `validation_fraction` not None.
  Reasoning: `validation_fraction` is just ignored and will never raise. On top, `None` already has another meaning.

Reply:
@adrinjalali Ok for you?

Reply:
I approved, but since the API changed since @adrinjalali's review, I'm going to not merge yet.